File: packages/tracelet_ios/ios/Classes/EventDispatcher.swift
import Flutter
import Foundation

/// Manages all 15 EventChannels, registering stream handlers and dispatching
/// events to Dart on the main thread.
final class EventDispatcher: NSObject {
    private var sinks: [String: FlutterEventSink] = [:]
    private var channels: [FlutterEventChannel] = []

    /// All event channel paths (must match Dart TraceletEvents constants).
    static let channelPaths: [String] = [
        "com.tracelet/events/location",
        "com.tracelet/events/motionchange",
        "com.tracelet/events/activitychange",
        "com.tracelet/events/providerchange",
        "com.tracelet/events/geofence",
        "com.tracelet/events/geofenceschange",
        "com.tracelet/events/heartbeat",
        "com.tracelet/events/http",
        "com.tracelet/events/schedule",
        "com.tracelet/events/powersavechange",
        "com.tracelet/events/connectivitychange",
        "com.tracelet/events/enabledchange",
        "com.tracelet/events/notificationaction",
        "com.tracelet/events/authorization",
        "com.tracelet/events/watchposition",
    ]

    func register(messenger: FlutterBinaryMessenger) {
        for path in EventDispatcher.channelPaths {
            let channel = FlutterEventChannel(name: path, binaryMessenger: messenger)
            channel.setStreamHandler(StreamHandler(path: path, dispatcher: self))
            channels.append(channel)
        }
    }

    func unregister() {
        for channel in channels {
            channel.setStreamHandler(nil)
        }
        channels.removeAll()
        sinks.removeAll()
    }

    // MARK: - Typed senders

    func sendLocation(_ data: [String: Any]) {
        send("com.tracelet/events/location", data: data)
    }

    func sendMotionChange(_ data: [String: Any]) {
        send("com.tracelet/events/motionchange", data: data)
    }

    func sendActivityChange(_ data: [String: Any]) {
        send("com.tracelet/events/activitychange", data: data)
    }

    func sendProviderChange(_ data: [String: Any]) {
        send("com.tracelet/events/providerchange", data: data)
    }

    func sendGeofence(_ data: [String: Any]) {
        send("com.tracelet/events/geofence", data: data)
    }

    func sendGeofencesChange(_ data: [String: Any]) {
        send("com.tracelet/events/geofenceschange", data: data)
    }

    func sendHeartbeat(_ data: [String: Any]) {
        send("com.tracelet/events/heartbeat", data: data)
    }

    func sendHttp(_ data: [String: Any]) {
        send("com.tracelet/events/http", data: data)
    }

    func sendSchedule(_ data: [String: Any]) {
        send("com.tracelet/events/schedule", data: data)
    }

    func sendPowerSaveChange(_ isPowerSave: Bool) {
        send("com.tracelet/events/powersavechange", data: isPowerSave)
    }

    func sendConnectivityChange(_ data: [String: Any]) {
        send("com.tracelet/events/connectivitychange", data: data)
    }

    func sendEnabledChange(_ enabled: Bool) {
        send("com.tracelet/events/enabledchange", data: enabled)
    }

    func sendNotificationAction(_ data: [String: Any]) {
        send("com.tracelet/events/notificationaction", data: data)
    }

    func sendAuthorization(_ data: [String: Any]) {
        send("com.tracelet/events/authorization", data: data)
    }

    func sendWatchPosition(_ data: [String: Any]) {
        send("com.tracelet/events/watchposition", data: data)
    }

    // MARK: - Core dispatch

    private func send(_ path: String, data: Any) {
        DispatchQueue.main.async { [weak self] in
            self?.sinks[path]?(data)
        }
    }

    // MARK: - Stream handler

    fileprivate func setSink(_ path: String, sink: FlutterEventSink?) {
        sinks[path] = sink
    }

    /// Internal stream handler that routes listen/cancel to the dispatcher.
    private class StreamHandler: NSObject, FlutterStreamHandler {
        let path: String
        weak var dispatcher: EventDispatcher?

        init(path: String, dispatcher: EventDispatcher) {
            self.path = path
            self.dispatcher = dispatcher
        }

        func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
            dispatcher?.setSink(path, sink: events)
            return nil
        }

        func onCancel(withArguments arguments: Any?) -> FlutterError? {
            dispatcher?.setSink(path, sink: nil)
            return nil
        }
    }
}

File: packages/tracelet_android/android/src/main/kotlin/com/tracelet/tracelet_android/EventDispatcher.kt
package com.tracelet.tracelet_android

import android.os.Handler
import android.os.Looper
import com.tracelet.core.TraceletEventSender
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

/**
 * Flutter-specific [TraceletEventSender] implementation using EventChannels.
 *
 * Manages 15 EventChannels and their EventSinks. All event dispatch
 * is marshalled to the main thread for Flutter platform channel safety.
 *
 * When no Dart UI listener is attached for a given event, the dispatcher
 * falls back to [headlessFallback] (if set) so that events can be routed
 * to a background Dart isolate via HeadlessTaskService.
 */
class EventDispatcher : TraceletEventSender {

    companion object {
        private const val BASE = "com.tracelet/events"
    }

    private val mainHandler = Handler(Looper.getMainLooper())

    /**
     * Optional headless fallback. When no EventSink exists for a given event,
     * the dispatcher calls this lambda with (eventName, eventData) so the
     * event can be forwarded to HeadlessTaskService.
     */
    var headlessFallback: ((eventName: String, data: Map<String, Any?>) -> Unit)? = null

    // EventChannel instances
    private val channels = mutableMapOf<String, EventChannel>()

    // Active EventSinks (null when no Dart listener)
    private val sinks = mutableMapOf<String, EventChannel.EventSink?>()

    // All event channel suffixes
    private val eventNames = listOf(
        "location",
        "motionchange",
        "activitychange",
        "providerchange",
        "geofence",
        "geofenceschange",
        "heartbeat",
        "http",
        "schedule",
        "powersavechange",
        "connectivitychange",
        "enabledchange",
        "notificationaction",
        "authorization",
        "watchposition",
    )

    /** Registers all EventChannels with the Flutter binary messenger. */
    fun register(messenger: BinaryMessenger) {
        for (name in eventNames) {
            val path = "$BASE/$name"
            val channel = EventChannel(messenger, path)
            channel.setStreamHandler(object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                    sinks[name] = events
                }
                override fun onCancel(arguments: Any?) {
                    sinks[name] = null
                }
            })
            channels[name] = channel
        }
    }

    /** Unregisters all EventChannels. */
    fun unregister() {
        for ((_, channel) in channels) {
            channel.setStreamHandler(null)
        }
        channels.clear()
        sinks.clear()
    }

    // ---------------------------------------------------------------------------
    // TraceletEventSender implementation (all marshal to main thread)
    // ---------------------------------------------------------------------------

    override fun sendLocation(data: Map<String, Any?>) = send("location", data)

    override fun sendMotionChange(data: Map<String, Any?>) = send("motionchange", data)

    override fun sendActivityChange(data: Map<String, Any?>) = send("activitychange", data)

    override fun sendProviderChange(data: Map<String, Any?>) = send("providerchange", data)

    override fun sendGeofence(data: Map<String, Any?>) = send("geofence", data)

    override fun sendGeofencesChange(data: Map<String, Any?>) = send("geofenceschange", data)

    override fun sendHeartbeat(data: Map<String, Any?>) = send("heartbeat", data)

    override fun sendHttp(data: Map<String, Any?>) = send("http", data)

    override fun sendSchedule(data: Map<String, Any?>) = send("schedule", data)

    override fun sendPowerSaveChange(isPowerSaveMode: Boolean) = send("powersavechange", isPowerSaveMode)

    override fun sendConnectivityChange(data: Map<String, Any?>) = send("connectivitychange", data)

    override fun sendEnabledChange(enabled: Boolean) = send("enabledchange", enabled)

    override fun sendNotificationAction(action: String) = send("notificationaction", action)

    override fun sendAuthorization(data: Map<String, Any?>) = send("authorization", data)

    override fun sendWatchPosition(data: Map<String, Any?>) = send("watchposition", data)

    /** Returns true if a listener is attached for the given event name. */
    override fun hasListener(eventName: String): Boolean = sinks[eventName] != null

    // ---------------------------------------------------------------------------
    // Private
    // ---------------------------------------------------------------------------

    private fun send(eventName: String, data: Any?) {
        val sink = sinks[eventName]
        if (sink != null) {
            if (Looper.myLooper() == Looper.getMainLooper()) {
                sink.success(data)
            } else {
                mainHandler.post { sink.success(data) }
            }
        } else {
            // No Dart UI listener — route to headless fallback if available.
            headlessFallback?.let { fallback ->
                @Suppress("UNCHECKED_CAST")
                val eventData = when (data) {
                    is Map<*, *> -> data as Map<String, Any?>
                    else -> mapOf("value" to data)
                }
                fallback(eventName, eventData)
            }
        }
    }
}
